File: Tests/Source/ExecutorTester.cpp
#include <optional>
#include <thread>
#include <vector>
#include <catch2/catch.hpp>
#include "Aspen/Constant.hpp"
#include "Aspen/Executor.hpp"
#include "Aspen/Lift.hpp"
#include "Aspen/None.hpp"
#include "Aspen/Queue.hpp"

using namespace Aspen;

TEST_CASE("test_run_until_none_empty", "[Executor]") {
  auto result = std::optional<int>();
  auto executor = Executor(
    lift([&] (const auto& value) {
      result = value;
    }, none<int>()));
  executor.run_until_none();
  REQUIRE(!result.has_value());
}

TEST_CASE("test_run_until_none_constant", "[Executor]") {
  auto result = std::optional<int>();
  auto executor = Executor(
    lift([&] (const auto& value) {
      result = value;
    }, constant(5)));
  executor.run_until_none();
  REQUIRE(result.has_value());
  REQUIRE(*result == 5);
}

TEST_CASE("test_run_until_complete", "[Executor]") {
  auto queue = Queue<int>();
  auto results = std::vector<int>();
  auto executor = Executor(
    lift([&] (const auto& value) {
      results.push_back(value);
    }, &queue));
  auto executor_thread = std::thread([&] {
    executor.run_until_complete();
  });
  queue.push(10);
  queue.push(20);
  queue.push(30);
  queue.push(40);
  queue.set_complete(100);
  executor_thread.join();
  REQUIRE(results.size() == 5);
  REQUIRE(results == std::vector{10, 20, 30, 40, 100});
}

File: Source/PythonTests/executor_tester.py
import threading
import unittest

from fixtures import aspen


class TestExecutor(unittest.TestCase):
  def test_run_until_none(self):
    commits = []
    reactor = aspen.lift(lambda value: commits.append(value), aspen.constant(1))
    aspen.Executor(reactor).run_until_none()
    self.assertEqual(commits, [1])

  def test_aborting_before_a_run(self):
    commits = []
    reactor = aspen.lift(lambda value: commits.append(value), aspen.Cell(0))
    executor = aspen.Executor(reactor)
    executor.abort()
    executor.run_until_complete()
    self.assertEqual(commits, [])

  def test_aborting_a_run(self):
    executor = aspen.Executor(aspen.Cell(0))
    abort = executor.abort
    runner = threading.Thread(target=executor.run_until_complete, daemon=True)
    runner.start()
    abort()
    runner.join(timeout=10)
    self.assertFalse(runner.is_alive())


if __name__ == '__main__':
  unittest.main()
